cluster/src/main/scala/org/apache/spark/memory/SnappyUnifiedMemoryManager.scala

```diff
@@ -562,26 +562,20 @@ class SnappyUnifiedMemoryManager private[memory](
   }
 
   override def releaseStorageMemoryForObject(objectName: String,
-                                             numBytes: Long,
-                                             memoryMode: MemoryMode): Unit = synchronized {
+      numBytes: Long,
+      memoryMode: MemoryMode): Unit = synchronized {
     logDebug(s"releasing $managerId memory for $objectName = $numBytes")
     val key = objectName -> memoryMode
-    if (memoryForObject.addTo(key, -numBytes) >= 0L) {
-      super.releaseStorageMemory(numBytes, memoryMode)
-    } else {
-      // objectName was not present
-      memoryForObject.removeLong(key)
+    super.releaseStorageMemory(numBytes, memoryMode)
+    if (memoryForObject.containsKey(key)) {
+      if (memoryForObject.addTo(key, -numBytes) == numBytes) {
+        memoryForObject.removeLong(key)
+      }
```

Comment:
I think original code was correct and more efficient. It avoids two lookup into the map for most cases. Something like below with your updated logic to always invoke super:

```scala
  super.releaseStorageMemory(numBytes, memoryMode)
  val originalNumBytes = memoryForObject.addTo(key, -numBytes)
  if (memoryForObject.addTo(key, -numBytes) <= numBytes) {
    // key not present or has gone down to zero
    memoryForObject.removeLong(key)
  }
```